File: code/cpp/2196.cpp
// https://leetcode.com/problems/create-binary-tree-from-descriptions/

class Solution {
public:
    TreeNode* getNode(auto& nodes, int val) {
        TreeNode* res;
        if (nodes.contains(val))
            res = nodes[val];
        else {
            res = new TreeNode(val);
            nodes[val] = res;
        }
        return res;
    }

    TreeNode* createBinaryTree(vector<vector<int>>& descriptions) {
        unordered_map<int, TreeNode*> nodes;
        unordered_set<int> children;

        for (const auto& description : descriptions) {
            int parent = description[0], 
                child = description[1];
            bool isLeft = description[2];
            TreeNode* parent_node = getNode(nodes, parent);
            children.insert(child);
            if (isLeft)
                parent_node->left = getNode(nodes, child);
            else
                parent_node->right = getNode(nodes, child);
        }
        for (auto& node : nodes)
            if (!children.contains(node.first))
                return node.second;
        return nullptr;
    }
};

File: code/python/2196.py
# https://leetcode.com/problems/create-binary-tree-from-descriptions/

class Solution:
    def createBinaryTree(self, descriptions: List[List[int]]) -> Optional[TreeNode]:
        def get_node(val) -> TreeNode:
            if val in nodes:
                res = nodes[val]
            else:
                res = TreeNode(val)
                nodes[val] = res
            return res
        
        nodes = {}
        children = set()
        for [parent, child, is_left] in descriptions:
            parent_node = get_node(parent)
            children.add(child)
            if is_left:
                parent_node.left = get_node(child)
            else:
                parent_node.right = get_node(child)

        for node in nodes.values():
            if node.val not in children:
                return node
